common/cmake/devices.cmake

```diff
@@ -929,6 +929,189 @@ function(ADD_FPGA_TARGET_BOARDS)
+  set(options)
+  set(oneValueArgs NAME USE_FASM OUT_LOCAL_REL)
+  set(multiValueArgs INCLUDED_TARGETS)
+  cmake_parse_arguments(
+    ADD_BITSTREAM_TARGET
+    "${options}"
+    "${oneValueArgs}"
+    "${multiValueArgs}"
+    ${ARGN}
+  )
+
+
+  set(NAME ${ADD_BITSTREAM_TARGET_NAME})
+  set(INCLUDED_TARGETS ${ADD_BITSTREAM_TARGET_INCLUDED_TARGETS})
+  set(USE_FASM ${ADD_BITSTREAM_TARGET_USE_FASM})
+  set(OUT_LOCAL_REL ${ADD_BITSTREAM_TARGET_OUT_LOCAL_REL})
+
+  if("${USE_FASM}" STREQUAL "")
+    set(USE_FASM TRUE)
+  endif()
+  # Generate bitstream
+  # -------------------------------------------------------------------------
+  set(ALL_OUT_FASM "")
+  if(${USE_FASM})
+    foreach(TARGET ${INCLUDED_TARGETS})
+      get_target_property_required(BOARD ${TARGET} BOARD)
+      get_target_property_required(DEVICE ${BOARD} DEVICE)
+      get_target_property_required(DEVICE_TYPE ${DEVICE} DEVICE_TYPE)
+      get_target_property(USE_OVERLAY ${DEVICE_TYPE} USE_OVERLAY)
+      get_target_property_required(FASM ${TARGET} FASM)
+      if ("${FASM_TO_BIT_DEPS}" STREQUAL "FASM_TO_BIT_DEPS-NOTFOUND")
+        set(FASM_TO_BIT_DEPS "")
+      endif()
+
+      append_file_location(ALL_OUT_FASM ${FASM})
+      append_file_dependency(ALL_OUT_FASM_DEPS ${FASM})
+
+      if(USE_OVERLAY)
+        if (NOT MAIN_TARGET)
+          set(MAIN_TARGET ${TARGET})
+        else()
+          message(FATAL_ERROR "More than one overlay device for ${BOARD}")
+        endif()
+      endif()
+    endforeach()
+
+    if (NOT MAIN_TARGET)
+      list(LENGTH ${INCLUDED_TARGETS} TARGETS_LENGTH)
+      if(NOT ${TARGETS_LENGTH} EQUAL 0)
+        message(FATAL_ERROR "Multiple devices but no overlay for ${BOARD}")
+      endif()
+      set(MAIN_TARGET ${INCLUDED_TARGETS})
+    endif()
+
+    get_target_property_required(MAIN_TARGET_BOARD ${MAIN_TARGET} BOARD)
+    get_target_property_required(MAIN_TARGET_DEVICE ${MAIN_TARGET_BOARD} DEVICE)
+    get_target_property_required(PACKAGE ${MAIN_TARGET_BOARD} PACKAGE)
+    get_target_property_required(PYTHON3 env PYTHON3)
+
+    get_target_property(FASM_TO_BIT_EXTRA_ARGS ${MAIN_TARGET_BOARD} FASM_TO_BIT_EXTRA_ARGS)
+    if ("${FASM_TO_BIT_EXTRA_ARGS}" STREQUAL "FASM_TO_BIT_EXTRA_ARGS-NOTFOUND")
+      set(FASM_TO_BIT_EXTRA_ARGS "")
+    endif()
+
+    get_target_property_required(TOP ${MAIN_TARGET} TOP)
+    get_target_property_required(ARCH ${MAIN_TARGET_DEVICE} ARCH)
+    get_target_property_required(BITSTREAM_EXTENSION ${ARCH} BITSTREAM_EXTENSION)
+    get_target_property_required(FASM_TO_BIT ${ARCH} FASM_TO_BIT)
+    get_target_property_required(FASM_TO_BIT_CMD ${ARCH} FASM_TO_BIT_CMD)
+    get_target_property(FASM_TO_BIT_DEPS ${ARCH} FASM_TO_BIT_DEPS)
+
+    if(NOT OUT_LOCAL_REL)
+      set(FQDN ${ARCH}-${DEVICE_TYPE}-${DEVICE}-${PACKAGE})
+      set(OUT_LOCAL_REL ${NAME}/${FQDN})
+      set(OUT_LOCAL ${CMAKE_CURRENT_BINARY_DIR}/${OUT_LOCAL_REL})
+      set(MAIN_TARGET ${NAME})
+      add_custom_target(${NAME})
+    else()
```

Comment:
This seems wrong?  `OUT_LOCAL_REL` is empty?  Is this the intended?

Reply:
I set `OUT_LOCAL_REL` one line up.  This is for the case where a bit stream target is being explicitly defined (for merged fasm bit streams).